systemverilog-plugin/UhdmAst.cc

```diff
@@ -2264,16 +2264,93 @@ void UhdmAst::process_cont_assign()
 
 void UhdmAst::process_assignment()
 {
+    const uhdm_handle *const handle = (const uhdm_handle *)obj_h;
+    const UHDM::BaseClass *const object = (const UHDM::BaseClass *)handle->object;
     auto type = vpi_get(vpiBlocking, obj_h) == 1 ? AST::AST_ASSIGN_EQ : AST::AST_ASSIGN_LE;
+    bool simple_assign = true;
+    bool shift_unsigned = false;
+    int op_type = vpi_get(vpiOpType, obj_h);
+    AST::AstNodeType node_type;
     current_node = make_ast_node(type);
+
+    if (op_type && op_type != 82) {
+        simple_assign = false;
+        visit_one_to_one({vpiLhs}, obj_h, [&](AST::AstNode *node) {
+            if (node) {
+                if (node->type == AST::AST_PARAMETER || node->type == AST::AST_LOCALPARAM) {
+                    node->type = AST::AST_IDENTIFIER;
+                }
+                current_node->children.push_back(node);
+            }
+        });
+        switch (op_type) {
+        case vpiSubOp:
+            node_type = AST::AST_SUB;
+            break;
+        case vpiDivOp:
+            node_type = AST::AST_DIV;
+            break;
+        case vpiModOp:
+            node_type = AST::AST_MOD;
+            break;
+        case vpiLShiftOp:
+            node_type = AST::AST_SHIFT_LEFT;
+            shift_unsigned = true;
+            break;
+        case vpiRShiftOp:
+            node_type = AST::AST_SHIFT_RIGHT;
+            shift_unsigned = true;
+            break;
+        case vpiAddOp:
+            node_type = AST::AST_ADD;
+            break;
+        case vpiMultOp:
+            node_type = AST::AST_MUL;
+            break;
+        case vpiBitAndOp:
+            node_type = AST::AST_BIT_AND;
+            break;
+        case vpiBitOrOp:
+            node_type = AST::AST_BIT_OR;
+            break;
+        case vpiBitXorOp:
+            node_type = AST::AST_BIT_XOR;
+            break;
+        case vpiArithLShiftOp:
+            node_type = AST::AST_SHIFT_SLEFT;
+            shift_unsigned = true;
+            break;
+        case vpiArithRShiftOp:
+            node_type = AST::AST_SHIFT_SRIGHT;
+            shift_unsigned = true;
+            break;
+        default:
+            delete current_node;
+            current_node = nullptr;
+            report_error("%s:%d: Encountered unhandled compound assignment with operation type %d\n", object->VpiFile().c_str(), object->VpiLineNo(),
+                         op_type);
+            return;
+        }
+        current_node->children.push_back(make_ast_node(node_type));
+    }
+
     visit_one_to_one({vpiLhs, vpiRhs}, obj_h, [&](AST::AstNode *node) {
         if (node) {
             if (node->type == AST::AST_PARAMETER || node->type == AST::AST_LOCALPARAM) {
                 node->type = AST::AST_IDENTIFIER;
             }
-            current_node->children.push_back(node);
+            if (!simple_assign) {
```

Comment:
In general we can always visit LHS and RHS first, then assemble the RHS depending on operation type (for simple assignment passing it directly, for others assembling the operation).

Reply:
That looks much better, thanks!